packages/core/src/modules/credentials/CredentialsApi.ts

```diff
@@ -0,0 +1,626 @@
+
+import { CredentialState } from './models/CredentialState'
+import { RevocationNotificationService } from './protocol/revocation-notification/services'
+import { V1CredentialService } from './protocol/v1/V1CredentialService'
+import { V2CredentialService } from './protocol/v2/V2CredentialService'
+import { CredentialRepository } from './repository/CredentialRepository'
+
+export interface CredentialsApi<CFs extends CredentialFormat[], CSs extends CredentialService<CFs>[]> {
+  // Proposal methods
+  proposeCredential(options: ProposeCredentialOptions<CFs, CSs>): Promise<CredentialExchangeRecord>
+  acceptProposal(options: AcceptProposalOptions<CFs>): Promise<CredentialExchangeRecord>
+  negotiateProposal(options: NegotiateProposalOptions<CFs>): Promise<CredentialExchangeRecord>
+
+  // Offer methods
+  offerCredential(options: OfferCredentialOptions<CFs, CSs>): Promise<CredentialExchangeRecord>
+  acceptOffer(options: AcceptOfferOptions<CFs>): Promise<CredentialExchangeRecord>
+  declineOffer(credentialRecordId: string): Promise<CredentialExchangeRecord>
+  negotiateOffer(options: NegotiateOfferOptions<CFs>): Promise<CredentialExchangeRecord>
+  // out of band
+  createOffer(options: CreateOfferOptions<CFs, CSs>): Promise<{
+    message: AgentMessage
+    credentialRecord: CredentialExchangeRecord
+  }>
+  // Request
+  // This is for beginning the exchange with a request (no proposal or offer). Only possible
+  // (currently) with W3C. We will not implement this in phase I
+  // requestCredential(credentialOptions: RequestCredentialOptions): Promise<CredentialExchangeRecord>
+
+  // when the issuer accepts the request he issues the credential to the holder
+  acceptRequest(options: AcceptRequestOptions<CFs>): Promise<CredentialExchangeRecord>
+
+  // Credential
+  acceptCredential(options: AcceptCredentialOptions): Promise<CredentialExchangeRecord>
+  sendProblemReport(options: SendProblemReportOptions): Promise<CredentialExchangeRecord>
+
+  // Record Methods
+  getAll(): Promise<CredentialExchangeRecord[]>
+  getById(credentialRecordId: string): Promise<CredentialExchangeRecord>
+  findById(credentialRecordId: string): Promise<CredentialExchangeRecord | null>
+  deleteById(credentialRecordId: string, options?: DeleteCredentialOptions): Promise<void>
+  getFormatData(credentialRecordId: string): Promise<GetFormatDataReturn<CFs>>
+
+  // DidComm Message Records
+  findProposalMessage(credentialExchangeId: string): Promise<FindProposalMessageReturn<CSs>>
+  findOfferMessage(credentialExchangeId: string): Promise<FindOfferMessageReturn<CSs>>
+  findRequestMessage(credentialExchangeId: string): Promise<FindRequestMessageReturn<CSs>>
+  findCredentialMessage(credentialExchangeId: string): Promise<FindCredentialMessageReturn<CSs>>
+}
+
+@injectable()
+export class CredentialsApi<
+  CFs extends CredentialFormat[] = [IndyCredentialFormat],
+  CSs extends CredentialService<CFs>[] = [V1CredentialService, V2CredentialService<CFs>]
```

Comment:
Will these kinds of generics only be used with the `credentialsApi`? I am worried about an api class with many generics and it becomes very difficult to comprehend what is going on.

Reply:
It will be used for protocols where this is needed. So I suspect it will only be credentials and proofs. Understand the note on generics being tricky. I'm not sure however how to make it both generic and simple